src/global/pidfile.cc

```diff
@@ -31,68 +31,131 @@
 
 #define dout_prefix *_dout
 
-static char pid_file[PATH_MAX] = "";
+struct pidfh {
+        int     pf_fd;
+        char    pf_path[PATH_MAX + 1];
+        dev_t   pf_dev;
+        ino_t   pf_ino;
+ 
+        pidfh() : pf_fd(-1), pf_dev(0), pf_ino(0) {
+                  memset(pf_path, 0, sizeof(pf_path));
+        }
+       
+        void close() {
+                pf_fd = -1;
+                pf_path[0] = '\0';
+                pf_dev = 0;
+                pf_ino = 0;
+        }
+};
+static struct pidfh pfh;
 
-int pidfile_write(const md_config_t *conf)
-{
-  int ret, fd;
+static int pidfile_verify() {
+     struct stat sb;
 
-  if (conf->pid_file.empty()) {
-    return pidfile_remove();
-  }
-  snprintf(pid_file, PATH_MAX, "%s", conf->pid_file.c_str());
+     if (pfh.pf_fd == -1)
+            return -EINVAL;
+     /*
+      * Check remembered descriptor
+      */ 
+    if (fstat(pfh.pf_fd, &sb) == -1)
+            return -errno;
+    if (sb.st_dev != pfh.pf_dev || sb.st_ino != pfh.pf_ino)
+           return -ESTALE;
+    return 0;
+}
 
-  fd = TEMP_FAILURE_RETRY(::open(pid_file,
-				 O_CREAT|O_TRUNC|O_WRONLY, 0644));
-  if (fd < 0) {
-    int err = errno;
-    derr << "write_pid_file: failed to open pid file '"
-	 << pid_file << "': " << cpp_strerror(err) << dendl;
-    return err;
-  }
+int pidfile_write()
+{
+  int ret;
+  if (!pfh.pf_path[0])
+      return 0;
 
-  char buf[20];
+  if ((ret=pidfile_verify()) < 0) {
+    return ret;
+  }
+  char buf[32];
   int len = snprintf(buf, sizeof(buf), "%d\n", getpid());
-  ret = safe_write(fd, buf, len);
+  ret = safe_write(pfh.pf_fd, buf, len);
   if (ret < 0) {
     derr << "write_pid_file: failed to write to pid file '"
-	 << pid_file << "': " << cpp_strerror(ret) << dendl;
-    VOID_TEMP_FAILURE_RETRY(::close(fd));
+	 << pfh.pf_path << "': " << cpp_strerror(ret) << dendl;
```

Comment:
@xiexingguo  all clean work will be done by pidfile_remove, which is called when daemon shutdown. so it seems not to cause fd leak, am i right?